notebooks/query-dependent link analysis.ipynb

```diff
@@ -0,0 +1,1064 @@
+    "\n",
+    "# Draw graph\n",
+    "showGraph(g)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4c35a7ff",
+   "metadata": {},
+   "source": [
+    "## Exploring the Start a Business subgraph"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "99b3668d",
+   "metadata": {},
+   "source": [
+    "Summary: \n",
+    "- Subgraph components = 274, 16, 3, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1\n",
+    "- Max shortest path length = 17\n",
+    "- Centrality metrics provide an estimate of node importance, pages such as \n",
+    "  '/get-information-about-a-company' and '/running-a-limited-company'\n",
+    "- Nodes with user movement between = 186\n",
+    "- User movement weight ranges from 6 to 34837"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "90df4bb9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Which nodes are in the list but not in the graph? \n",
+    "## Answer: two pages in the list no longer exist, so are not in the graph\n",
+    "\n",
+    "# List of node property 'name'\n",
+    "nodes_graph = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    nodes_graph.append(v['properties']['name'])\n",
+    "\n",
+    "# Which nodes are in sab_pages but not in the graph\n",
+    "[node for node in sab_pages if node not in nodes_graph]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d7a80d8f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Subgraph components: [274, 16, 3, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
+    "nx.is_weakly_connected(g)\n",
+    "[len(c) for c in sorted(nx.weakly_connected_components(g), key=len, reverse=True)]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a93854ed",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "list(nx.weakly_connected_components(g))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a090b285",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Max shortest path length: 17\n",
+    "max([max(j.values()) for (i,j) in nx.shortest_path_length(g)]) # find the largest diameter amongst all components within g"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "823e754d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Between centrality: the number of shortest paths that pass through the node - a 'bridge' between nodes\n",
+    "g_di = nx.DiGraph(g)\n",
+    "g_di_between = nx.betweenness_centrality(g_di)\n",
+    "g_between = dict(sorted(g_di_between.items(), key=operator.itemgetter(1),reverse=True))\n",
+    "\n",
+    "\n",
+    "node_names = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    node_names.append(k)\n",
+    "    node_names.append(v['properties']['name']) # append cid id and name    \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
```

Comment:
Can this all be re written as:

node_names_dict = {k:v['properties']['name'] for k,v in g.nodes(data=True)}

??

Reply:
Thank you, refactored in commit a3a488951396dda849cfc7c6de72e78d07499605